<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Fundamentals (Good)</span></div>

# What to expect in this chapter

# 1 There is more to if

In [1]:
name = 'Batman'

if name == 'Batman':
    print('Hello Batman!')
elif name == 'Robin':
    print('Hello Sidekick1!')
elif name == 'Tom':
    print('Hello Sidekick2!')
elif name == 'Jerry':
    print('Hello Sidekick3!')
elif name == 'Perry':
    print('Hello Sidekick4!')
else:
    print('Hello World!')

Hello Batman!


# 2 Asking questions

In [8]:
fruits = ['apple', 'banana', 'pineapple', 'jackfruit']
vegetables = ['celery', 'potato', 'broccoli', 'kale']
('apple' in fruits) or ('kale' in vegetables)


True

In [6]:
fruits = ['apple', 'banana', 'pineapple', 'jackfruit']
vegetables = ['celery', 'potato', 'broccoli', 'kale']

('apple' in fruits) and ('kale' in vegetables)


False

In [7]:
fruits = ['apple', 'banana', 'pineapple', 'jackfruit']
vegetables = ['celery', 'potato', 'broccoli', 'kale']

'kale' not in vegetables

False

In [9]:
'app' in 'apple'

True

In [11]:
'apple' > 'oranges'
# The answer is false, as the programme sees each letter as a code/number and compares the numbers between both words

False

## 2.1 Asking Math questions

# 3 Python stores information in different formats or types

In [12]:
x = int(1.234)
print(x, type(x))

1 <class 'int'>


In [15]:
x = str(1.234)
print(f"{x=}", type(x))

x='1.234' <class 'str'>


In [18]:
x = float(1.234)
print(x, type(x))


1.234 <class 'float'>


In [16]:
x = complex(1.234)
print(f"{x=}",type(x))

x=(1.234+0j) <class 'complex'>


In [22]:
import numpy as np
np.float16([10])

array([10.], dtype=float16)

In [23]:
np.float32([10])

array([10.], dtype=float32)

In [24]:
np.float64([10])

array([10.])

# 4 Never compare floats directly

In [25]:
a = 0.1
a3 = 0.3
a * 3 == a3

False

In [26]:
f'{0.3:.17f}'

'0.29999999999999999'

Internally, the computer is unable to store an infinite number of zeros in 0.3000~


In [27]:
abs(a3 - 3*a) < 1E-5


True

In [29]:
np.isclose(a3,3*a)

np.True_

Use on of the **above two** methods to compare float variables

## 4.1 The Problem

## 4.2 A solution

# 5 Combining English and variables

## 5.1 Structure of f-strings

In [30]:
name = "Batman"
print(f"Hello {name}!")

Hello Batman!


In [31]:
name = "Batman"
print(f"Hello {name.upper()}!")

Hello BATMAN!


In [32]:
text = 'Bruce Wayne is Batman.'
print(f'{text}')
print(f'{text:>30}')      # A block of 30 characters;

Bruce Wayne is Batman.
        Bruce Wayne is Batman.


In [33]:
print(f'The cube of pi to 6 decimal places is {np.pi**3:05.6f}')

The cube of pi to 6 decimal places is 31.006277


In [34]:
print(f'The cube of pi to 6 decimal places is {np.pi**3:.6f}')

The cube of pi to 6 decimal places is 31.006277


In [37]:
print(f'The cube of pi to 6 decimal places is {int(np.pi**3):05d}')

The cube of pi to 6 decimal places is 00031


# 6 Escape sequences

In [39]:
print('Line 1\n:Line 2\tLine 3')

Line 1
:Line 2	Line 3


In [40]:
print('Line 1\n:Lin\\e 2\tLine 3') # To put a slash (\) in a string, you must put 2 slashes together

Line 1
:Lin\e 2	Line 3


## 6.1 Self-documenting f-strings

# 7 Computers read = from Right to Left!

In [43]:
x = 40
y = x + 2
y

42

In [44]:
y = 40
y = y + 2
print(y)

42


In [45]:
y = 40
y = y + 2
y
y = 40
y += 2    # Same as y = y + 2
y

42

# 8 Shorter and Cleaner Code

# 9 Python can be a prima-donna.

# 10 Best Practices for Scientific Computing

# 11 Looking for help

## References

## Footnotes